src/System.Private.CoreLib/shared/System/SpanHelpers.Byte.cs

```diff
@@ -255,7 +266,91 @@ public static unsafe int IndexOf(ref byte searchSpace, byte value, int length)
+                    }
+                    else
+                    {
+                        return (int)(byte*)index + TrailingZeroCountFallback(matches);
+                    }
+                }
+
+                nLength = (IntPtr)((length - (int)(byte*)index) & ~(Vector128<byte>.Count - 1));
+                if ((byte*)nLength > (byte*)index)
+                {
+                    Vector128<byte> comparison128 = Vector128.Create(value);
+
+                    Vector128<byte> vSearch = Unsafe.ReadUnaligned<Vector128<byte>>(ref Unsafe.AddByteOffset(ref searchSpace, index));
+                    int matches = Sse2.MoveMask(Sse2.CompareEqual(comparison128, vSearch));
+                    if (matches == 0)
+                    {
+                        index += Vector128<byte>.Count;
+                    }
+                    // Find offset of first match
+                    else if (Bmi1.IsSupported)
+                    {
+                        return ((int)(byte*)index) + (int)Bmi1.TrailingZeroCount((uint)matches);
+                    }
+                    else
+                    {
+                        return (int)(byte*)index + TrailingZeroCountFallback(matches);
+                    }
+                }
+
+                if ((int)(byte*)index < length)
+                {
+                    nLength = (IntPtr)(length - (int)(byte*)index);
+                    goto SequentialScan;
+                }
+            }
+            else if (!Avx2.IsSupported && Sse2.IsSupported && ((int)(byte*)index < length))
+            {
+                nLength = (IntPtr)((length - (int)(byte*)index) & ~(Vector128<byte>.Count - 1));
+
+                Vector128<byte> vComparison = Vector128.Create(value);
+                while ((byte*)nLength > (byte*)index)
+                {
+                    Vector128<byte> vSearch = Unsafe.ReadUnaligned<Vector128<byte>>(ref Unsafe.AddByteOffset(ref searchSpace, index));
+                    int matches = Sse2.MoveMask(Sse2.CompareEqual(vComparison, vSearch));
+                    if (matches == 0)
+                    {
+                        index += Vector128<byte>.Count;
+                        continue;
+                    }
+                    // Find offset of first match
+                    else if (Bmi1.IsSupported)
+                    {
+                        return ((int)(byte*)index) + (int)Bmi1.TrailingZeroCount((uint)matches);
+                    }
+                    else
+                    {
+                        return (int)(byte*)index + TrailingZeroCountFallback(matches);
+                    }
+                }
+
+                if ((int)(byte*)index < length)
+                {
+                    nLength = (IntPtr)(length - (int)(byte*)index);
+                    goto SequentialScan;
+                }
+            }
+            else if (!Avx2.IsSupported && !Sse2.IsSupported && Vector.IsHardwareAccelerated && ((int)(byte*)index < length))
```

Comment:
Following up from your comment in https://github.com/dotnet/coreclr/pull/22019#discussion_r249292983 the Jit doesn't seem to branch eliminate without the not Supported tests; the fact its an else doesn't seem to do it 😢

Reply:
Sounds like a JIT bug.

@AndyAyersMS Is it something easy to fix?

Reply:
For clarity
```csharp
else if (Vector.IsHardwareAccelerated && ((int)(byte*)index < length))
```
Will include the block (as `Vector.IsHardwareAccelerated` is true); however the earlier `Avx2.IsSupported` true in the else if chain should eliminate it.

OTOH this will drop the block
```csharp
else if (!Avx2.IsSupported && !Sse2.IsSupported && Vector.IsHardwareAccelerated && ((int)(byte*)index < length))
```

Not sure if it was the case for the opening tests; maybe the complexity inside the `if`, `else if` bodies pushed it over a limit?

Reply:
I may not be looking at the right example, but 
```
if ((Avx2.IsSupported || Sse2.IsSupported) && length >= Vector128<byte>.Count * 2)
{
    ...
}
else if (Vector.IsHardwareAccelerated && length >= Vector<byte>.Count * 2)
{
    ...
}
```
When Avx2 is supported, the first clause reduces to `length > 32`, and the second to `length > 64`, so I don't see how the first makes the second redundant.

A `Vector128` is always 128 bits.

Reply:
Ah... good point!

So it would need to be 
```
if ((Avx2.IsSupported || Sse2.IsSupported))
{
   if (length >= Vector128<byte>.Count * 2) 
   {
       ...
   }
}
else if (Vector.IsHardwareAccelerated)
{
    if (length >= Vector<byte>.Count * 2)
    {
        ...
    }
}
```
Not sure the extra indent flows better than the current version? Might keep it the same

Reply:
Ah, I see - sorry for the false alarm.

@benaadams Would it be better to write it like this to avoid the redundant `IsSupported` conditions?

```csharp
if (Avx2.IsSupported)
{
    if ((int)(byte*)index < length)
    {
...
    }
}
else if (Sse2.IsSupported)
{
    if ((int)(byte*)index < length)
    {
...
    }
}
else if (Vector.IsHardwareAccelerated)
{
    if ((int)(byte*)index < length)
    {
...
    }
}
```

Reply:
Ok, will do

Reply:
Done